packages/curve-ui-kit/src/features/connect-wallet/lib/ConnectionContext.tsx

```diff
@@ -0,0 +1,178 @@
+  TChainId extends number,
+  TLib extends { chainId: TChainId; signerAddress?: string },
+>({
+  hydrate,
+  initLib,
+  chainId,
+  onChainUnavailable,
+  children,
+}: {
+  hydrate: (newLib: TLib, prevLib: TLib | null, wallet: Wallet | null) => Promise<void>
+  initLib: (chainId: TChainId, wallet: Wallet | null) => Promise<TLib>
+  chainId: TChainId
+  onChainUnavailable: ([unsupportedChainId, walletChainId]: [TChainId, TChainId]) => void
+  children: ReactNode
+}) => {
+  const [connectState, setConnectState] = useState<ConnectState>({ status: LOADING })
+  const { wallet, connect } = useWallet()
+  const [_, setChain] = useSetChain()
+
+  const walletRef = useRef(wallet) // use ref to avoid re-rendering on wallet change
+  walletRef.current = wallet
+
+  const walletChainId = getWalletChainId(wallet)
+  const walletSignerAddress = getWalletSignerAddress(wallet)
+  const walletLabel = wallet?.label
+
+  useEffect(() => {
+    const abortController = new AbortController()
+    const walletName = getFromLocalStorage<string>(WalletNameStorageKey) // todo: we might not need the walletName at all in useWallet
+
+    const connectWalletStage = async () => {
+      let wallet = walletRef.current
+      if (walletLabel != walletName) {
+        const connectOptions = { ...(walletName && { autoSelect: { label: walletName, disableModals: true } }) }
+        ;[wallet] = await withTimeout(connect(connectOptions))
+        setLocalStorage(WalletNameStorageKey, wallet?.label ?? null)
+      }
+      return wallet
+    }
+
+    const initApp = async () => {
+      try {
+        setConnectState({ status: LOADING, stage: CONNECT_WALLET })
+        const wallet = await connectWalletStage()
+        if (abortController.signal.aborted) return
+
+        if (walletChainId && walletChainId !== chainId) {
+          setConnectState({ status: LOADING, stage: SWITCH_NETWORK })
+          if (!(await setChain({ chainId: ethers.toQuantity(chainId) }))) {
+            if (abortController.signal.aborted) return
+            setConnectState({ status: FAILURE, stage: SWITCH_NETWORK })
+            onChainUnavailable([chainId, walletChainId as TChainId])
+          }
+        }
+
+        const prevLib = libRef.get<TLib>()
+        if (!libRef.get() || walletSignerAddress != prevLib?.signerAddress) {
+          if (abortController.signal.aborted) return
+          setConnectState({ status: LOADING, stage: CONNECT_API })
+          const newLib = await initLib(chainId, wallet)
+          if (abortController.signal.aborted) return
+          libRef.set(newLib)
+        }
+
+        if (abortController.signal.aborted) return
+        setConnectState({ status: SUCCESS, stage: HYDRATE })
+        await hydrate(libRef.require<TLib>(), prevLib, wallet)
+
+        if (abortController.signal.aborted) return
+        setConnectState({ status: SUCCESS })
+      } catch (error) {
+        console.error(error)
+        if (abortController.signal.aborted) return
```

Comment:
It is the correct way for the time being but damn all these abort checks lol

Reply:
yes, it's super ugly. Ultimately we could set up a pipeline with steps and iterate over them